Fix drop order of tls and statics and avoid double panics on deadlock

Pointerbender opened the pull request:
Hello! I ran into some of the same issues as mentioned in #152, #179 and #291 and developed a fix for all of those issues. In a nutshell, the changes are:

- clarify the documentation of the mock `lazy_static` implementation regarding its `Drop` implementation
- ensure that global statics are dropped after thread locals
- ensure that `loom` does not panic inside a panic when dropping thread locals on the main thread
- add two regression tests for #152
- ensure that thread locals are destructed when a spawned thread panics
- ensure that loom doesn't double panic on deadlock
- add a test to verify that thread locals are properly destructed on spawned threads
- sprinkle comments generously everywhere to explain the code changes

This is my first time contributing to `loom`, please let me know if I can somehow improve this PR or if additional information is needed. Thank you!

Pointerbender committed:
- clarify documentation of mock `lazy_static` implementation
- ensure that global statics are dropped after thread locals
- ensure that `loom` does not panic inside a panic
- add two regression tests for #152

Pointerbender committed:
- ensure that thread locals are destructed when a spawned thread panics
- ensure that loom doesn't double panic on deadlock
- add a test to verify that thread locals are properly destructed on spawned threads

Pointerbender commented:
Hello! Just checking in to see if there is anything I can do to help shepherd this PR along. :) I'd be happy to provide extra details or hop on chat/zoom to review together if that saves some time. No rush, though!

hawkw commented:
Thank you for working on this, this is great! I'll try to give it a review at some point soon, although I imagine @carllerche will want to as well.

adrianheine commented on src/model.rs:
```diff
@@ -182,14 +182,37 @@ impl Builder {
+                /// and this then would in turn panic from the method
+                /// [`Lazy::get`](crate::lazy_static::Lazy), which
+                /// causes a panic inside a panic, which in turn causes
+                /// Rust to abort, triggering a segfault in `loom`.
```
These should be regular comments, not doc comments.

Pointerbender replied:
These doc comments intentionally apply to the `struct PanicGuard;`

adrianheine commented on src/rt/mod.rs:
```diff
@@ -73,8 +73,27 @@ where
+        /// when this happens, we still want to ensure that
+        /// thread locals are destructed. therefore, we set
+        /// up a guard that is dropped as part of the unwind
+        /// logic when `f` panics.
```
These should be regular comments, not doc comments.

Pointerbender replied:
These doc comments intentionally apply to the `struct PanicGuard;`

Pointerbender commented:
I've addressed the PR comments left by @adrianheine (thanks for the review!), are there any next steps left before this could potentially be merged? :) Please let me know if I can be of any further assistance. Thank you!

Hywan commented:
Hey :-),
Any news about this PR?
